ibm_mq/datadog_checks/ibm_mq/ibm_mq.py

```diff
@@ -79,29 +79,35 @@ def check(self, instance):
 
     def discover_queues(self, queue_manager, config):
         queues = []
-        if config.auto_discover_queues:
+        has_patterns = len(config.queue_patterns) > 0
+        if config.auto_discover_queues or has_patterns:
             queues.extend(self._discover_queues(queue_manager, '*'))
-
-        if len(config.queue_patterns) > 0:
-            for regex in config.queue_patterns:
-                queues.extend(self._discover_queues(queue_manager, regex))
+            if has_patterns:
+                keep_queues = []
+                for queue_pattern in config.queue_patterns:
+                    queue_pattern_re = re.compile(queue_pattern)
+                    for queue in queues:
+                        if bool(queue_pattern_re.match(queue)):
+                            keep_queues.append(queue)
+                queues = keep_queues
 
         config.add_queues(queues)
 
-    def _discover_queues(self, queue_manager, regex):
-        args = {pymqi.CMQC.MQCA_Q_NAME: ensure_bytes(regex), pymqi.CMQC.MQIA_Q_TYPE: pymqi.CMQC.MQQT_MODEL}
+    def _discover_queues(self, queue_manager, mq_pattern_filter):
         queues = []
 
-        try:
-            pcf = pymqi.PCFExecute(queue_manager)
-            response = pcf.MQCMD_INQUIRE_Q(args)
-        except pymqi.MQMIError as e:
-            self.warning("Error getting queue stats: {}".format(e))
-        else:
-            for queue_info in response:
-                queue = queue_info[pymqi.CMQC.MQCA_Q_NAME]
-                queue = queue.strip()
-                queues.append(queue)
+        for queue_type in self.SUPPORTED_QUEUE_TYPES:
+            args = {pymqi.CMQC.MQCA_Q_NAME: ensure_bytes(mq_pattern_filter), pymqi.CMQC.MQIA_Q_TYPE: queue_type}
+            try:
+                pcf = pymqi.PCFExecute(queue_manager)
+                response = pcf.MQCMD_INQUIRE_Q(args)
+            except pymqi.MQMIError as e:
+                self.warning("Error discovering queue: {}".format(e))
+            else:
+                for queue_info in response:
+                    queue = queue_info[pymqi.CMQC.MQCA_Q_NAME]
+                    if queue_info[pymqi.CMQC.MQIA_DEFINITION_TYPE] == pymqi.CMQC.MQQDT_PREDEFINED:
```

Comment:
> Filtering out of 'system' queues is also updated to look at the DEFTYPE attribute to determine a 'system' queue in lieu of the hard-coded list found in the existing version of config.py -- this filter is now based on whether a queue's DEFTYPE attribute is PREDEFINED (if so, consider it, else ignore).

Hum. Does this condition filter the system queues ? `if queue_info[pymqi.CMQC.MQIA_DEFINITION_TYPE] == pymqi.CMQC.MQQDT_PREDEFINED:`

Seems that it might filter out all MODEL queues. According to the doc:

> Predefined queues cannot be created from model queues.